### ISSUE TITLE
Method to broadcast parameters/buffers of DDP model

### ISSUE THREAD

**Issue by pietern** [pytorch/pytorch#30718]
## 🚀 Feature

Add DDP method to allow a user to broadcast parameters and/or buffers manually.

## Motivation

We only have the constructor kwarg `broadcast_buffers` at the moment (defaults to `True`, many folks set it to `False` for performance reasons). This flag toggles between never broadcasting buffers or always broadcasting buffers. There may be some schedule that is more optimal for the user (e.g. once per epoch) that is impossible to encode in DDP as is. Therefore, we should add a function to broadcast the buffers (and while we're at it, the parameters), so the user can call it whenever.

## Pitch

```python
def broadcast_parameters(self, root=0):
  # ...
def broadcast_buffers(self, root=0):
  # ... 
```

We'll need to rename the boolean member field and move existing functionality around.

## Additional context

The broadcast of model parameters that is done from the constructor can directly use the proposed function as well. The broadcast done today is a "coalesced" version, where first a subset of tensors is concatenated in a single buffer before running the broadcast, to make it more bandwidth-optimal.

**Comment by edpizzi** [comment]
To give a concrete use case -- we want to synchronize BatchNorm stats across GPUs so that eval reflects a consistent set of BN stats, and eval metrics are reproducible from the saved snapshot (which typically reflects GPU 0's BN stats).

Using `broadcast_buffers` achieves this, at the cost of an extra broadcast every training iteration, whereas for this use case, synchronization is only needed once per epoch, between training and eval phases. Allowing an explicit sync once per epoch solves this problem with trivial performance overhead.

**Comment by zhaojuanmao** [comment]
@edpizzi are you blocked by this feature? or you've already have some workarounds now?

**Comment by zhaojuanmao** [comment]
@pietern are you suggesting followings?
1) rename "broadcast_buffers" flag to be something like "require_forward_buffers_sync",  but it will have back compatibility issue, right? could we keep the flag name unchanged?
2) split _sync_params() into two public APIs:  broadcast_parameters() and broadcast_buffers()

**Comment by edpizzi** [comment]
@zhaojuanmao I'm not blocked, thanks. And your suggestion looks good to me, provided renaming the `broadcast_buffers` attribute is acceptable.
